Handle an optional symengine and sympy in qpy

mtreinish opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

This commit changes the API for qpy.load() to make it explicit that symengine may not be installed in the future but is a requirement for deserializing certain payloads that were generated using symengine. We're expecting to drop symengine as a requirement with #13278 which should improve Qiskit's installation story and also will facilitate adding a C api for circuits. But the qpy v10, v11, and v13 have a hard requirement on symengine (specifically 0.11.0 and 0.13.0) to be able to deserialize `ParameterExpression` if the symbolic encoding was set to use it. Since the function needs to support being able to deserialize these payloads symengine will need to be installed. This commit adds the document and release note to indicate this change in behavior.

Similarly, `qpy.dump()` will have issues if symengine is not used anymore for the internal representation for QPY 10, 11, and 12 of an expression relies on symengine. So we won't necessarily be able to generate a symengine expression losslessly or without any overhead so trying to support this in a model without symengine installed isn't a tenable. This commit raises the minimum compatibility version so symengine (or sympy) isn't needed to generate the payload anymore.

### Details and comments

mtreinish committed:
Document that symengine is optional for qpy.load

This commit changes the API for qpy.load() to make it explicit that
symengine may not be installed in the future but is a requirement for
deserializing certain payloads that were generated using symengine.
We're expecting to drop symengine as a requirement with #13278 which
should improve Qiskit's installation story and also will facilitate
adding a C api for circuits. But the qpy v10, v11, and v13 have a hard
requirement on symengine (specifically 0.11.0 and 0.13.0) to be able to
deserialize `ParameterExpression` if the symbolic encoding was set to
use it. Since the function needs to support being able to deserialize
these payloads symengine will need to be installed. This commit adds the
document and release note to indicate this change in behavior.

jakelishman reviewed (commented):
Do you potentially want to put in the necessary `HAS_SYMENGINE` optional and the `@require_in_call` decorators with this PR, so that the whole principle is in place?

mtreinish committed:
Raise QPY compatibility version to remove symengine and sympy dependency

This commit raises the qpy compatibility version to QPY format version
13. As the larger PR is planning for a world without symengine installed
by default this becomes a problem for the generation side too. This
becomes a blocker for using QPY < 13 in the future so this commit opts
to just raise the minimum verison to get ahead of any potential issues.

mtreinish committed:
Add HAS_SYMENGINE check on load function

mtreinish committed:
Merge branch 'main' into optional-qpy-symengine

mtreinish commented on test/python/qpy/test_circuit_load_from_qpy.py:
```diff
@@ -209,52 +209,6 @@ def test_no_register(self, opt_level):
 class TestVersionArg(QpyCircuitTestCase):
     """Test explicitly setting a qpy version in dump()."""
 
-    def test_custom_gate_name_overlap_persists_with_minimum_version(self):
```
This test was removed because it was explicitly testing the bug fixed in QPY 11 was present if we specifically set the QPY version to 10. Since the bug is still fixed in QPY 13, there isn't anything to test anymore since we can't generate a buggy payload.

jakelishman commented on qiskit/qpy/interface.py:
```diff
@@ -134,11 +134,9 @@ def dump(
-            but not supported in all platforms. This flag only has an effect if the emitted QPY format
-            version is 10, 11, or 12. For QPY format version >= 13 (which is the default starting in
-            Qiskit 1.3.0) this flag is no longer used.
+        use_symengine: This flag is no longer used by QPY versions supported by this function and
+            will have no impact on the generated QPY payload except to set a field in a QPY v13 file
+            header which is unused.
```
It was potentially a mistake for us not to update `symbolic_encoding` in version 13 to be some typecode that's independent of sympy and symengine.  I guess there's not much to be done about it now.

mtreinish replied:
Yeah, it was an oversight, we were a bit rushed trying to get QPY13 in for 1.3. It's something we could potentially fix in v14, but it's a bit tight timewise to slip that in to v14. But we also missed the deprecation of the kwarg here so we'd be stuck with an arg that has no impact either way.

jakelishman replied:
I'm not all that beaten up about the kwarg tbh, it's more the payload.  Right now, with QPY 13 already having shipped with a meaningless field there, I'm not sure how much purpose there is in changing it in QPY 14.

jakelishman commented on qiskit/qpy/common.py:
```diff
@@ -311,13 +308,19 @@ def mapping_from_binary(binary_data, deserializer, **kwargs):
 
 
-def load_symengine_payload(payload: bytes) -> symengine.Expr:
+@HAS_SYMENGINE.require_in_call
```
Since this is an internal function, we might want to give `require_in_call` a string argument like `require_in_call("QPY version 10--12 payloads with symengine parameters")`  to explain what the failure is?  Otherwise it'll be pretty esoteric for the user, like "symengine is required to use 'load_symengine_payload'" or something.

mtreinish replied:
Added in: https://github.com/Qiskit/qiskit/pull/13917/commits/8b9ec52f47498bc901abf91991ea6b12775a8590

jakelishman commented on qiskit/qpy/binary_io/circuits.py:
```diff
@@ -755,13 +755,12 @@ def _write_instruction(
         # in circuits with multiple ucr*_dg gates.
-        elif instruction.operation.name in {"ucrx_dg", "ucry_dg", "ucrz_dg"}:
+        if instruction.operation.name in {"ucrx_dg", "ucry_dg", "ucrz_dg"}:
             gate_class_name = f"{gate_class_name}_{uuid.uuid4()}"
```
I think we're duplicating the uuid append to the ucr* gates now?

mtreinish replied:
Ah good catch, I was just mechanically updating the logic to remove the version branch.

mtreinish replied:
Fixed in: https://github.com/Qiskit/qiskit/pull/13917/commits/36c2623d8abb98bec5b347668d2f0aa92748cad9

jakelishman reviewed (commented):
For most of this PR I'm happy, but I'm a bit concerned that we didn't issue deprecation warnings in Qiskit 1.4 about the withdrawal of support for writing QPY versions 10--12.  How painful would it be to continue to support them, and document that the optional is required to support write out of those QPY versions with symengine/sympy?

mtreinish commented:
> For most of this PR I'm happy, but I'm a bit concerned that we didn't issue deprecation warnings in Qiskit 1.4 about the  withdrawal of support for writing QPY versions 10--12. How painful would it be to continue to support them, and document that the optional is required to support write out of those QPY versions with symengine/sympy?

The way this is implemented in the #13278 currently is that it basically stringifies the inner rust expression object, runs it through a regex find replace to update the syntax, then passes that string to sympy to create a new expression object. If the sympy object is then either used verbatim if sympy encoding is used or converted to a symengine expression and that is encoded. It seems to work for all our current tests that use an explicit version, but it seems overly fragile for a core piece of qiskit.

FWIW, we did document this case in the [stability guidelines](https://docs.quantum.ibm.com/open-source/qiskit-sdk-version-strategy#upgrade-strategy) and the dump docstring to be only guaranteed to work for the n-1 release, so I  think we have a leg to stand on here. Obviously it would have been much better if we did warn about it in <= 1.4.0, but that ship has sailed.

mtreinish committed:
Merge remote-tracking branch 'origin/main' into optional-qpy-symengine

mtreinish committed:
Correct logic for ucr*_dg gates name changing

mtreinish committed:
Add feature string to optional decorator

mtreinish committed:
Prepare for an optional sympy

In addition to making symengine optional #13278 should also enable us to
make sympy optional. It would only be potentially needed for loading QPY
payloads that were generated using sympy encoding, but also some
visualization functions. This commit lays the groundwork for removing it
as a dependency by outlining the API changes for the functions that will
optionally depend on sympy and raise an exception if it's not installed.
It won't be possible to trigger the exception with sympy in the
requirements list, but when we do remove it the API is prepared for it.

mtreinish committed:
Fix release note categorization

mtreinish committed:
Merge remote-tracking branch 'origin/main' into optional-qpy-symengine

mtreinish committed:
Add missing optional on template substitution

kevinhartman commented on qiskit/qpy/interface.py:
```diff
@@ -252,9 +250,13 @@ def load(
-        QiskitError: if ``file_obj`` is not a valid QPY file.
+        QiskitError: if ``file_obj`` is not a valid QPY file
+        TypeError: When invalid data type is loaded.
+        MissingOptionalLibraryError: If the ``symengine`` engine library is
```
🚂

kevinhartman reviewed (approved):
This looks good to me, though I'll avoid adding it to the queue in case anyone else is still looking.

Personally, I don't see much issue with dropping support for writing V12 and older, since this is squarely allowed in our stability guidelines.

mtreinish commented:
Given the time crunch I've enqueued this for merge
